File: test/headway.test.mjs
// Trains cannot occupy the same block.
//
// Positions are estimated per train, from its own timetable and its own delay,
// so nothing stops two of them being drawn on the same piece of track: a fast
// train catching a slower one is shown closing right up to it. In reality it is
// held a block short and slowed — a regular sight between Bordeaux and Dax.
//
// Note what this is not: SNCF does not publish signal positions (the dataset
// named for them is a computer-vision corpus, no coordinates), so this is not a
// claim about which signal is at danger. It is the weaker and still true claim
// that a train cannot be where another one already is.

import { test } from 'node:test';
import assert from 'node:assert/strict';
import { fileURLToPath } from 'node:url';
import path from 'node:path';

const ROOT = path.join(path.dirname(fileURLToPath(import.meta.url)), '..');
const { analyseTraffic, headingGap } = await import(path.join(ROOT, 'dist-server/server/Headway.js'));
const { BlockIndex, blockLengthFor } = await import(path.join(ROOT, 'dist-server/server/Blocks.js'));

/** A train running between stops, `km` north of a base point. */
function running(number, line, km, progress, bearing = 0) {
  return {
    number,
    line,
    position: {
      basis: 'between',
      lat: 44 + km / 111.32,
      lon: -0.5,
      bearing,
      progress,
      speedKmh: 160,
    },
  };
}

/** Two kilometres of block everywhere. */
const twoKm = () => 2000;

test('a train closing on the one ahead is held back', () => {
  // 8542 is 800 m ahead of 8540 on the same line, both northbound.
  const t = analyseTraffic([running('8540', 'L1', 0, 0.4), running('8542', 'L1', 0.8, 0.5)], twoKm);

  const follower = t.get('8540');
  assert.equal(follower.ahead, '8542');
  assert.equal(follower.aspect, 'semaphore', 'the block ahead is occupied');
  assert.ok(follower.pushedM > 1000 && follower.pushedM < 1300, `pushed ${follower.pushedM} m`);

  // The leader has nothing in front of it.
  assert.equal(t.get('8542').aspect, 'libre');
  assert.equal(t.get('8542').pushedM, undefined);
});

test('one block clear reads as a warning, not as clear', () => {
  // 3 km behind on 2 km blocks: the section ahead is empty, but the one after
  // it is not, so the signal in front of this train warns.
  const t = analyseTraffic([running('8540', 'L1', 0, 0.4), running('8542', 'L1', 3, 0.5)], twoKm);
  assert.equal(t.get('8540').aspect, 'avertissement');
  assert.equal(t.get('8540').pushedM, undefined, 'a clear block needs no correction');
});

test('two blocks clear reads as clear', () => {
  const t = analyseTraffic([running('8540', 'L1', 0, 0.4), running('8542', 'L1', 6, 0.5)], twoKm);
  assert.equal(t.get('8540').aspect, 'libre');
  assert.equal(t.get('8540').pushedM, undefined);
});

test('the train in front is never held by the one behind', () => {
  const t = analyseTraffic([running('8540', 'L1', 0, 0.4), running('8542', 'L1', 0.5, 0.6)], twoKm);
  assert.equal(t.get('8542').pushedM, undefined, 'the leader must not be pushed by its follower');
  assert.equal(t.get('8542').ahead, undefined);
});

test('trains going opposite ways do not constrain each other', () => {
  // Two tracks, one each way: they pass, they do not queue.
  const t = analyseTraffic(
    [running('8540', 'L1', 0, 0.4, 0), running('9001', 'L1', 0.5, 0.5, 180)],
    twoKm,
  );
  for (const v of t.values()) assert.equal(v.pushedM, undefined, 'passing is not following');
});

test('trains on different lines do not constrain each other', () => {
  const t = analyseTraffic([running('8540', 'L1', 0, 0.4), running('7000', 'L2', 0.4, 0.5)], twoKm);
  for (const v of t.values()) assert.equal(v.ahead, undefined);
});

test('a train standing in a station is not held', () => {
  const stopped = running('8540', 'L1', 0, 0.4);
  stopped.position.basis = 'at_station';
  const t = analyseTraffic([stopped, running('8542', 'L1', 0.5, 0.5)], twoKm);
  assert.ok(!t.has('8540'), 'where the timetable puts it is where it is');
});

test('the tightest constraint wins when several are ahead', () => {
  const held = analyseTraffic(
    [
      running('8540', 'L1', 0, 0.3),
      running('8542', 'L1', 1.5, 0.5),
      running('8544', 'L1', 0.6, 0.4), // closer, so this is the binding one
    ],
    twoKm,
  );
  assert.equal(held.get('8540').ahead, '8544');
});

test('trains far apart on a long line are not paired', () => {
  const t = analyseTraffic([running('8540', 'L1', 0, 0.4), running('8542', 'L1', 120, 0.5)], twoKm);
  assert.equal(t.get('8540').ahead, undefined, '120 km apart is not following');
});

test('nothing is held where there is no block working', () => {
  // "Sans cantonnement" means no spacing to enforce; inventing one would be
  // worse than leaving the estimate alone.
  const t = analyseTraffic([running('8540', 'L1', 0, 0.4), running('8542', 'L1', 0.2, 0.5)], () => 0);
  assert.equal(t.get('8540').pushedM, undefined);
  assert.equal(t.get('8540').aspect, 'inconnu', 'no signalling modelled means no claim');
});

test('headings wrap correctly around north', () => {
  assert.equal(headingGap(350, 10), 20);
  assert.equal(headingGap(10, 350), 20);
  assert.equal(headingGap(0, 180), 180);
});

// ── block lengths from the published working mode ────────────────────────────

test('each published mode maps to a plausible spacing', () => {
  assert.equal(blockLengthFor('Block automatique lumineux'), 1800);
  assert.equal(blockLengthFor('Block automatique lumineux de voie banalisée'), 1800);
  assert.equal(blockLengthFor('Transmission voie-machine 430'), 1500);
  assert.equal(blockLengthFor('Sans cantonnement'), 0);
  assert.ok(blockLengthFor('Block automatique à permissivité restreinte de voie unique') > 5000);
  assert.ok(blockLengthFor('Cantonnement téléphonique de voie unique') > 10000);
});

test('an unrecognised mode is not guessed at', () => {
  assert.equal(blockLengthFor('_Autre'), null);
  assert.equal(blockLengthFor(undefined), null);
});

test('the index answers by line and kilometre post', () => {
  const idx = new BlockIndex([
    { code_ligne: '570000', libelle: 'Block automatique lumineux', pkd: '010+000', pkf: '050+000' },
    {
      code_ligne: '570000',
      libelle: 'Cantonnement téléphonique de voie unique',
      pkd: '050+000',
      pkf: '090+000',
    },
  ]);

  assert.equal(idx.spacingFor('570000', 20, 30), 1800, 'inside the lit-block section');
  assert.equal(idx.spacingFor('570000', 60, 70), 20000, 'inside the telephone section');
  // Spanning both: the longer block binds, since a train must clear it all.
  assert.equal(idx.spacingFor('570000', 40, 60), 20000);
});

test('an unknown line falls back rather than failing', () => {
  const idx = new BlockIndex([]);
  assert.ok(idx.spacingFor('999999', 1, 2) > 0, 'a default block, not zero');
  assert.ok(idx.spacingFor(undefined, null, null) > 0);
});


// ── slowing for a signal ─────────────────────────────────────────────────────

const { approachSpeed } = await import(path.join(ROOT, 'dist-server/server/Headway.js'));

test('the approach follows the parabolic law', () => {
  // v = sqrt(2·a·d) at 0.5 m/s²: 1800 m gives 42 m/s, about 152 km/h.
  assert.ok(Math.abs(approachSpeed(1800, 300) - 152) < 2, `${approachSpeed(1800, 300)}`);
  // Four times the distance, twice the speed — the signature of the law.
  const near = approachSpeed(500, 300);
  const far = approachSpeed(2000, 300);
  assert.ok(Math.abs(far / near - 2) < 0.02, `${near} -> ${far}`);
});

test('a train slows the closer it gets', () => {
  let last = Infinity;
  for (const d of [4000, 3000, 2000, 1000, 500, 200, 50]) {
    const v = approachSpeed(d, 300);
    assert.ok(v < last, `${d} m allowed ${v}, not less than ${last}`);
    last = v;
  }
});

test('at the signal itself the train is stopped', () => {
  assert.equal(approachSpeed(0, 300), 0);
  assert.equal(approachSpeed(-500, 300), 0, 'past it, and certainly stopped');
});

test('a signal never licenses a train to go faster', () => {
  // Ten kilometres of clear track does not entitle a 90 km/h train to 500.
  assert.equal(approachSpeed(10_000, 90), 90);
});

test('coming off a restriction, speed returns quickly then gently', () => {
  // A square root, so for each further kilometre of clearance the train gains
  // less than it did for the last. Compared over equal steps — over doublings
  // the absolute gain grows, which is the curve behaving, not misbehaving.
  const early = approachSpeed(2000, 300) - approachSpeed(1000, 300);
  const late = approachSpeed(4000, 300) - approachSpeed(3000, 300);
  assert.ok(early > late, `gained ${early.toFixed(0)} then ${late.toFixed(0)} km/h per km`);
});

test('an approach to a restriction rather than a stop keeps that speed', () => {
  // Slowing to 80 rather than to a stand.
  assert.ok(approachSpeed(0, 300, 80) === 80, 'at the board, exactly the restriction');
  assert.ok(approachSpeed(1000, 300, 80) > 80, 'and more before reaching it');
});

test('the reported allowance appears only when it bites', () => {
  // Two blocks clear at 160 km/h: the braking curve permits more than the
  // train is doing, so there is nothing to report.
  const far = analyseTraffic(
    [running('8540', 'L1', 0, 0.4), running('8542', 'L1', 8, 0.5)],
    twoKm,
  );
  assert.equal(far.get('8540').allowedKmh, undefined);

  // Closing to within a block and a half, it does bite.
  const near = analyseTraffic(
    [running('8540', 'L1', 0, 0.4), running('8542', 'L1', 2.4, 0.5)],
    twoKm,
  );
  const t = near.get('8540');
  assert.ok(t.allowedKmh < 160, `allowed ${t.allowedKmh} of 160`);
  assert.ok(t.allowedKmh > 0);
});


test('two halves of a coupled train do not hold each other back', () => {
  // Seen in production: 12177 and 5537 are one train, Strasbourg to Nice,
  // drawn at the same point because they are in the same place. Without this
  // one of them was pushed a block back for running into itself.
  const a = running('12177', 'L1', 0, 0.5);
  const b = running('5537', 'L1', 0, 0.5);
  a.coupledWith = ['5537'];
  b.coupledWith = ['12177'];

  const t = analyseTraffic([a, b], twoKm);
  assert.equal(t.get('12177').pushedM, undefined, 'a train cannot follow itself');
  assert.equal(t.get('12177').ahead, undefined);
  assert.equal(t.get('5537').pushedM, undefined);
});

test('the exclusion holds even if only one side records the coupling', () => {
  const a = running('12177', 'L1', 0, 0.4);
  const b = running('5537', 'L1', 0.3, 0.6);
  a.coupledWith = ['5537'];
  // b.coupledWith deliberately left unset.
  const t = analyseTraffic([a, b], twoKm);
  assert.equal(t.get('12177').pushedM, undefined);
});

test('an ordinary follower is still held when a coupled set is nearby', () => {
  const a = running('12177', 'L1', 0, 0.5);
  const b = running('5537', 'L1', 0, 0.5);
  a.coupledWith = ['5537'];
  b.coupledWith = ['12177'];
  // A genuine third train closing on them.
  const c = running('9999', 'L1', -0.8, 0.3);

  const t = analyseTraffic([a, b, c], twoKm);
  assert.ok(t.get('9999').pushedM > 0, 'a real follower must still be held');
  assert.ok(['12177', '5537'].includes(t.get('9999').ahead));
});


// ── single track ─────────────────────────────────────────────────────────────

const SINGLE = () => ({ single: true, tracks: 1 });
const DOUBLE = () => ({ single: false, tracks: 2 });

/** Same as `running`, but pointed the other way. */
function facing(number, line, km, progress) {
  return running(number, line, km, progress, 180);
}

test('on single track, trains meeting head-on constrain each other', () => {
  // Neither is "ahead" of the other in any useful sense — they simply cannot
  // both continue.
  const t = analyseTraffic(
    [running('8540', 'L1', 0, 0.4), facing('9001', 'L1', 1.5, 0.5)],
    twoKm,
    undefined,
    SINGLE,
  );
  assert.equal(t.get('8540').aspect, 'semaphore');
  assert.equal(t.get('8540').opposing, true);
  assert.equal(t.get('8540').ahead, '9001');
  assert.equal(t.get('9001').opposing, true, 'both are constrained, not just one');
});

test('on double track, the same pair is ignored', () => {
  // Two tracks, one each way: they pass, which is the whole point of two.
  const t = analyseTraffic(
    [running('8540', 'L1', 0, 0.4), facing('9001', 'L1', 1.5, 0.5)],
    twoKm,
    undefined,
    DOUBLE,
  );
  assert.equal(t.get('8540').opposing, undefined);
  assert.equal(t.get('8540').pushedM, undefined);
});

test('neither opposing train is moved, because which one waits is unknowable', () => {
  // One of them is standing in a passing loop. The timetable does not say
  // which, and inventing an answer would put a train where it is not.
  const t = analyseTraffic(
    [running('8540', 'L1', 0, 0.4), facing('9001', 'L1', 1, 0.5)],
    twoKm,
    undefined,
    SINGLE,
  );
  assert.equal(t.get('8540').pushedM, undefined);
  assert.equal(t.get('9001').pushedM, undefined);
});

test('an opposing train slows for the meeting point', () => {
  const t = analyseTraffic(
    [running('8540', 'L1', 0, 0.4), facing('9001', 'L1', 0.6, 0.5)],
    twoKm,
    undefined,
    SINGLE,
  );
  const a = t.get('8540');
  assert.ok(a.allowedKmh < 160, `allowed ${a.allowedKmh} of 160`);
  assert.ok(a.allowedKmh >= 0);
});

test('a distant opposing train on single track is not a constraint', () => {
  // Twenty kilometres apart with two-kilometre blocks: there is a loop between
  // them, and treating that as a conflict would freeze half the network.
  const t = analyseTraffic(
    [running('8540', 'L1', 0, 0.4), facing('9001', 'L1', 20, 0.5)],
    twoKm,
    undefined,
    SINGLE,
  );
  assert.equal(t.get('8540').opposing, undefined);
});

test('following still works on single track', () => {
  // Same direction, single track: the ordinary following rule applies.
  const t = analyseTraffic(
    [running('8540', 'L1', 0, 0.4), running('8542', 'L1', 0.8, 0.5)],
    twoKm,
    undefined,
    SINGLE,
  );
  assert.equal(t.get('8540').ahead, '8542');
  assert.ok(t.get('8540').pushedM > 0, 'a follower is still held back');
  assert.equal(t.get('8540').opposing, undefined);
});

test('coupled halves are excluded on single track too', () => {
  const a = running('12177', 'L1', 0, 0.5);
  const b = running('5537', 'L1', 0, 0.5);
  a.coupledWith = ['5537'];
  b.coupledWith = ['12177'];
  const t = analyseTraffic([a, b], twoKm, undefined, SINGLE);
  assert.equal(t.get('12177').opposing, undefined);
  assert.equal(t.get('12177').pushedM, undefined);
});

File: server/internal/headway/headway_test.go
package headway

import (
	"math"
	"testing"

	"traincon/internal/train"
)

// blocks of a fixed length, which is what an unsignalled estimate gives.
func fixedSpacing(m float64) SpacingFunc {
	return func(_, _ float64) float64 { return m }
}

// at builds a follower running north at a given point, speed and progress.
func at(number string, lat, lon, bearing, kmh, progress float64) Follower {
	return Follower{
		Number: number,
		Line:   "L",
		Position: train.Position{
			Basis:    train.Between,
			Lat:      lat,
			Lon:      lon,
			Bearing:  bearing,
			SpeedKmh: kmh,
			Progress: progress,
		},
	}
}

func TestApproachSpeed(t *testing.T) {
	tests := []struct {
		name      string
		distanceM float64
		freeKmh   float64
		targetKmh float64
		want      float64
		tolerance float64
	}{
		// The worked example from the rule: 1 800 m from a stop signal is
		// about 150 km/h.
		{"1800 m from a red", 1800, 300, 0, 150, 3},
		// Right on top of it, nothing is permitted.
		{"at the signal", 0, 300, 0, 0, 0.001},
		// A restriction rather than a stop: the target speed carries through.
		{"approaching a 100 restriction", 0, 300, 100, 100, 0.001},
		// Far enough away that the train's own speed is the binding limit.
		{"miles off", 100000, 160, 0, 160, 0.001},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			got := ApproachSpeed(tc.distanceM, tc.freeKmh, tc.targetKmh)
			if math.Abs(got-tc.want) > tc.tolerance {
				t.Errorf("got %.1f km/h, want %.1f", got, tc.want)
			}
		})
	}
}

func TestApproachSpeedNeverLicencesGoingFaster(t *testing.T) {
	// A signal can restrain a train; it can never permit one to exceed the line
	// speed or its own timetable.
	for _, d := range []float64{0, 500, 5000, 50000} {
		if got := ApproachSpeed(d, 120, 0); got > 120 {
			t.Errorf("at %v m: %v km/h, above the free speed of 120", d, got)
		}
	}
}

func TestHeadingGap(t *testing.T) {
	tests := []struct{ a, b, want float64 }{
		{0, 0, 0},
		{0, 90, 90},
		{0, 180, 180},
		{10, 350, 20}, // across north
		{350, 10, 20}, // and back
		{0, 270, 90},  // the short way round
	}
	for _, tc := range tests {
		if got := headingGap(tc.a, tc.b); math.Abs(got-tc.want) > 1e-9 {
			t.Errorf("headingGap(%v,%v) = %v, want %v", tc.a, tc.b, got, tc.want)
		}
	}
}

func TestATrainWithNothingAheadRunsFree(t *testing.T) {
	got := Analyse([]Follower{at("1", 48, 2, 0, 160, 0.5)}, fixedSpacing(1500), nil, nil)
	if got["1"].Aspect != Libre {
		t.Errorf("aspect = %q, want libre", got["1"].Aspect)
	}
	if got["1"].Ahead != "" {
		t.Errorf("named %q ahead, but there is nothing there", got["1"].Ahead)
	}
}

func TestWithoutBlockWorkingNothingIsClaimed(t *testing.T) {
	// Calling a line clear whose signalling cannot be modelled would be a claim
	// too far.
	got := Analyse([]Follower{at("1", 48, 2, 0, 160, 0.5)}, fixedSpacing(0), nil, nil)
	if got["1"].Aspect != Inconnu {
		t.Errorf("aspect = %q, want inconnu", got["1"].Aspect)
	}
}

func TestAFollowerInsideABlockIsHeldBack(t *testing.T) {
	// Two trains 500 m apart on 1 500 m blocks: the follower is inside the
	// block ahead, so it is at a sémaphore and must be pushed back to the
	// boundary.
	const gapKm = 0.5
	trains := []Follower{
		at("behind", 48.0, 2.0, 0, 160, 0.4),
		at("ahead", 48.0+gapKm/111, 2.0, 0, 120, 0.6),
	}
	got := Analyse(trains, fixedSpacing(1500), nil, nil)

	held := got["behind"]
	if held.Aspect != Semaphore {
		t.Errorf("aspect = %q, want semaphore", held.Aspect)
	}
	if held.Ahead != "ahead" {
		t.Errorf("ahead = %q, want the train in front", held.Ahead)
	}
	if held.PushedM == nil || *held.PushedM <= 0 {
		t.Error("the follower was not pushed back out of the occupied block")
	}
	if held.AllowedKmh == nil || *held.AllowedKmh >= 160 {
		t.Errorf("allowed speed = %v, want a restriction below 160", held.AllowedKmh)
	}
	// The train in front is unconstrained.
	if got["ahead"].Aspect != Libre {
		t.Errorf("the leading train got %q, want libre", got["ahead"].Aspect)
	}
}

func TestAFollowerOneBlockBackSeesAWarning(t *testing.T) {
	// Between one and two blocks: the next signal is at danger, so this one
	// warns rather than stops.
	const gapKm = 2.0 // between 1 and 2 blocks of 1.5 km
	trains := []Follower{
		at("behind", 48.0, 2.0, 0, 160, 0.4),
		at("ahead", 48.0+gapKm/111, 2.0, 0, 120, 0.6),
	}
	got := Analyse(trains, fixedSpacing(1500), nil, nil)
	if got["behind"].Aspect != Avertissement {
		t.Errorf("aspect = %q, want avertissement", got["behind"].Aspect)
	}
	if got["behind"].PushedM != nil {
		t.Error("a train a full block back was pushed, but it is where it may be")
	}
}

func TestACoupledPortionIsNotFollowingItsOwnTwin(t *testing.T) {
	// Two portions joined together are one train carrying two numbers, drawn
	// at the same point because they are in the same place. 12177/5537,
	// Strasbourg to Nice, had one pushed a block back off the other.
	const gapKm = 0.4
	a := at("12177", 48.0, 2.0, 0, 160, 0.4)
	a.CoupledWith = []string{"5537"}
	b := at("5537", 48.0+gapKm/111, 2.0, 0, 160, 0.6)
	b.CoupledWith = []string{"12177"}

	got := Analyse([]Follower{a, b}, fixedSpacing(1500), nil, nil)
	for _, n := range []string{"12177", "5537"} {
		if got[n].PushedM != nil {
			t.Errorf("%s was pushed back off its own twin", n)
		}
		if got[n].Aspect != Libre {
			t.Errorf("%s got %q, want libre", n, got[n].Aspect)
		}
	}
}

func TestTrainsAtTheSamePointAreOneTrain(t *testing.T) {
	// Portions that run joined and split later share an origin rather than a
	// terminus, so the coupling detector misses them: 5500 to Metz and 12184 to
	// Strasbourg leave Montpellier at the same second, joined.
	trains := []Follower{
		at("5500", 48.0, 2.0, 0, 160, 0.4),
		at("12184", 48.0+0.05/111, 2.0, 0, 160, 0.6), // 50 m apart
	}
	got := Analyse(trains, fixedSpacing(1500), nil, nil)
	if got["5500"].PushedM != nil {
		t.Error("a train was pushed back off one at the same point")
	}
}

func TestATrainIsNotHeldByOneBehindIt(t *testing.T) {
	// Progress decides which is in front; a train behind constrains nothing.
	trains := []Follower{
		at("front", 48.0, 2.0, 0, 160, 0.8),
		at("back", 48.0+0.5/111, 2.0, 0, 160, 0.2), // nearer the pole, but behind
	}
	got := Analyse(trains, fixedSpacing(1500), nil, nil)
	if got["front"].Ahead != "" {
		t.Errorf("the leading train was held by %q", got["front"].Ahead)
	}
}

func TestTrainsGoingOppositeWaysDoNotFollowEachOther(t *testing.T) {
	trains := []Follower{
		at("north", 48.0, 2.0, 0, 160, 0.4),
		at("south", 48.0+0.5/111, 2.0, 180, 160, 0.6),
	}
	got := Analyse(trains, fixedSpacing(1500), nil, nil)
	if got["north"].Aspect != Libre {
		t.Errorf("aspect = %q, want libre — the other train is coming the other way", got["north"].Aspect)
	}
}

func TestOnSingleTrackAnOncomingTrainIsTheConstraint(t *testing.T) {
	// They cannot pass at all, so one is standing in a loop. Both are reported
	// and neither is moved: which one is waiting is not knowable from the
	// timetable.
	single := func(_, _ float64) Layout { return Layout{Single: true, Tracks: 1} }
	trains := []Follower{
		at("north", 48.0, 2.0, 0, 100, 0.4),
		at("south", 48.0+1.0/111, 2.0, 180, 100, 0.6),
	}
	got := Analyse(trains, fixedSpacing(1500), nil, single)

	for _, n := range []string{"north", "south"} {
		if !got[n].Opposing {
			t.Errorf("%s was not marked as facing an oncoming train", n)
		}
		if got[n].Aspect != Semaphore {
			t.Errorf("%s got %q, want semaphore", n, got[n].Aspect)
		}
		if got[n].PushedM != nil {
			t.Errorf("%s was moved, but which train is waiting is not knowable", n)
		}
	}
}

func TestRealSignalGeometryBeatsTheBlockEstimate(t *testing.T) {
	// Where the signalling layer knows where the signal is, that is where the
	// train would actually be brought to a stand.
	const gapKm = 2.0
	trains := []Follower{
		at("behind", 48.0, 2.0, 0, 200, 0.4),
		at("ahead", 48.0+gapKm/111, 2.0, 0, 120, 0.6),
	}
	sig := func(_, _, _ float64) (Signal, bool) {
		return Signal{M: 400, Kind: "carre"}, true
	}
	got := Analyse(trains, fixedSpacing(1500), sig, nil)

	if got["behind"].SignalM == nil || *got["behind"].SignalM != 400 {
		t.Errorf("signal distance = %v, want 400", got["behind"].SignalM)
	}
	if got["behind"].SignalKind != "carre" {
		t.Errorf("signal kind = %q, want carre", got["behind"].SignalKind)
	}
	// 400 m from a stand allows far less than the 1 250 m the block estimate
	// would have given.
	if got["behind"].AllowedKmh == nil || *got["behind"].AllowedKmh > 80 {
		t.Errorf("allowed speed = %v, want the tighter figure the real signal gives",
			got["behind"].AllowedKmh)
	}
}

func TestOnlyRunningTrainsAreConstrained(t *testing.T) {
	// One standing in a station is where the timetable says it is, and one that
	// has arrived is done.
	standing := at("standing", 48.0, 2.0, 0, 0, 0.4)
	standing.Position.Basis = train.AtStation
	trains := []Follower{standing, at("ahead", 48.0+0.5/111, 2.0, 0, 120, 0.6)}

	got := Analyse(trains, fixedSpacing(1500), nil, nil)
	if _, held := got["standing"]; held {
		t.Error("a train standing in a station was given traffic")
	}
}

func TestTrainsOnDifferentLinesDoNotInteract(t *testing.T) {
	a := at("1", 48.0, 2.0, 0, 160, 0.4)
	b := at("2", 48.0+0.5/111, 2.0, 0, 160, 0.6)
	b.Line = "another"
	got := Analyse([]Follower{a, b}, fixedSpacing(1500), nil, nil)
	if got["1"].Ahead != "" {
		t.Errorf("held by %q, which is on a different line", got["1"].Ahead)
	}
}
